crates/app/src/app.rs

```diff
@@ -0,0 +1,190 @@
+        eprintln!("failed to check repository existence: {}", e);
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Failed to check repository existence",
+        )
+    }
+    if !repos
+        .read()
+        .await
+        .contains(repo.clone())
+        .await
+        .map_err(err_map)?
+    {
+        Err((StatusCode::NOT_FOUND, "Repository does not exist"))
+    } else {
+        Ok(())
+    }
+}
+
+pub(crate) async fn assert_tag(
+    tags: Arc<TagStore>,
+    repo: repository::Name,
+    tag: tag::Name,
+) -> Result<(), (StatusCode, &'static str)> {
+    #[inline]
+    fn err_map(e: impl Display) -> (StatusCode, &'static str) {
+        eprintln!("failed to check tag existence: {}", e);
+        (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Failed to check tag existence",
+        )
+    }
+    if !tags
+        .read()
+        .await
+        .contains((repo.clone(), tag.clone()))
+        .await
+        .map_err(err_map)?
+    {
+        Err((StatusCode::NOT_FOUND, "Tag does not exist"))
+    } else {
+        Ok(())
+    }
+}
+
+/// Parses the URI of `req` and routes it to respective component.
+async fn handle(mut req: Request<Body>) -> impl IntoResponse {
+    let path = req.uri().path().strip_prefix('/').expect("invalid URI");
+    let (head, tail) = path
+        .split_once("/_")
+        .map(|(left, right)| (left, format!("_{}", right)))
+        .unwrap_or((path, "".into()));
+    if head.is_empty() {
+        return Err((
+            StatusCode::NOT_FOUND,
+            format!("Route `/{}` not found", path),
+        ));
+    }
+    let repo = head.parse::<repository::Name>().map_err(|e| {
+        (
+            StatusCode::BAD_REQUEST,
+            format!("Failed to parse repository name: {}", e),
+        )
+    })?;
+    assert_eq!(
+        req.extensions_mut().insert(repo),
+        None,
+        "duplicate repository name"
+    );
+
+    let mut tail = tail.split('/');
+    match (tail.next(), tail.next(), tail.next()) {
+        (None | Some(""), None, None) => match *req.method() {
+            Method::HEAD => Ok(repos::head.into_service().call(req).await.into_response()),
+            Method::GET => Ok(repos::get.into_service().call(req).await.into_response()),
+            Method::PUT => Ok(repos::put.into_service().call(req).await.into_response()),
+            _ => Err((
+                StatusCode::METHOD_NOT_ALLOWED,
+                "Method not allowed for repository endpoint".into(),
+            )),
+        },
+        (Some("_tag"), None, None) => match *req.method() {
+            Method::GET => Ok(tags::query.into_service().call(req).await.into_response()),
+            _ => Err((
+                StatusCode::METHOD_NOT_ALLOWED,
+                "Method not allowed for repository tag query endpoint".into(),
+            )),
+        },
+        (Some("_tag"), Some(tag), prop @ (None | Some("tree"))) => {
```

Comment:
What is this syntax with `@`?

Reply:
https://doc.rust-lang.org/rust-by-example/flow_control/match/binding.html